src/core/pointcloud/expression/qgspointcloudexpressionnodeimpl.cpp

```diff
@@ -0,0 +1,801 @@
+{
+  Q_UNUSED( parent )
+  Q_UNUSED( p )
+  return mValue;
+}
+
+QgsPointCloudExpressionNode::NodeType QgsPointCloudExpressionNodeLiteral::nodeType() const
+{
+  return ntLiteral;
+}
+
+bool QgsPointCloudExpressionNodeLiteral::prepareNode( QgsPointCloudExpression *parent, const QgsPointCloudBlock *block )
+{
+  Q_UNUSED( parent )
+  Q_UNUSED( block )
+  return true;
+}
+
+
+QString QgsPointCloudExpressionNodeLiteral::valueAsString() const
+{
+  return QString( "%1" ).arg( mValue );
+}
+
+QString QgsPointCloudExpressionNodeLiteral::dump() const
+{
+  return valueAsString();
+}
+
+QSet<QString> QgsPointCloudExpressionNodeLiteral::referencedAttributes() const
+{
+  return QSet<QString>();
+}
+
+QList<const QgsPointCloudExpressionNode *> QgsPointCloudExpressionNodeLiteral::nodes() const
+{
+  QList<const QgsPointCloudExpressionNode *> lst;
+  lst << this;
+  return lst;
+}
+
+QgsPointCloudExpressionNode *QgsPointCloudExpressionNodeLiteral::clone() const
+{
+  QgsPointCloudExpressionNodeLiteral *copy = new QgsPointCloudExpressionNodeLiteral( mValue );
+  cloneTo( copy );
+  return copy;
+}
+
+bool QgsPointCloudExpressionNodeLiteral::isStatic( QgsPointCloudExpression *parent, const QgsPointCloudBlock *block ) const
+{
+  Q_UNUSED( parent )
+  Q_UNUSED( block )
+  return true;
+}
+
+//
+
+double QgsPointCloudExpressionNodeAttributeRef::evalNode( QgsPointCloudExpression *parent, int p )
+{
+  Q_UNUSED( parent )
+
+  if ( !mAttribute || mOffset < 0 )
+  {
+    parent->setEvalErrorString( tr( "Attribute '%1' not found" ).arg( mName ) );
+    return std::numeric_limits<double>::quiet_NaN();
+  }
+
+  const char *data = mBlock->data();
+  const QgsPointCloudAttribute::DataType type = mAttribute->type();
+  const int offset = mBlock->attributes().pointRecordSize() * p + mOffset;
+
+  double val;
+  switch ( type )
+  {
+    case QgsPointCloudAttribute::UChar:
+    case QgsPointCloudAttribute::Char:
+      val = *( data + offset );
+      break;
+
+    case QgsPointCloudAttribute::UInt32:
+      val = *reinterpret_cast< const quint32 * >( data + offset );
+      break;
+    case QgsPointCloudAttribute::Int32:
+      val = *reinterpret_cast< const qint32 * >( data + offset );
+      break;
+
+    case QgsPointCloudAttribute::UInt64:
+      val = *reinterpret_cast< const quint64 * >( data + offset );
+      break;
+    case QgsPointCloudAttribute::Int64:
+      val = *reinterpret_cast< const qint64 * >( data + offset );
+      break;
+
+    case QgsPointCloudAttribute::Short:
+      val = *reinterpret_cast< const short * >( data + offset );
+      break;
+    case QgsPointCloudAttribute::UShort:
+      val = *reinterpret_cast< const unsigned short * >( data + offset );
+      break;
+
+    case QgsPointCloudAttribute::Float:
+      val = *reinterpret_cast< const float * >( data + offset );
+      break;
+    case QgsPointCloudAttribute::Double:
+      val = *reinterpret_cast< const double * >( data + offset );
+      break;
+  }
```

Comment:
what do you think about moving this code to a function in QgsPointCloudAttribute, e.g.
```
double convertValueToDouble( const char *ptr );
```

Reply:
ok. Should I make it static like `void getPointXYZ()` though?